python/sanlock.c

```diff
@@ -918,50 +915,103 @@ py_inq_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
+        convert_to_pybytes, &lockspace, &(ls->host_id), pypath_converter, &path,
+        &(ls->host_id_disk.offset), &unused, &wait)) {
+        goto finally;
+    }
+#endif
+    /* prepare sanlock names */
+    strncpy(ls->name, PyBytes_AsString(lockspace), SANLK_NAME_LEN);
+    strncpy(ls->host_id_disk.path, PyBytes_AsString(path), SANLK_PATH_LEN - 1);
+
+    /* prepare sanlock_rem_lockspace flags */
+    if (!wait) {
+        *flags |= SANLK_REM_ASYNC;
+    }
+
+    if (unused) {
+        *flags |= SANLK_REM_UNUSED;
+    }
+    rv = 1;
+
+finally:
+    Py_XDECREF(lockspace);
+    Py_XDECREF(path);
+    return rv;
+}
+
 /* rem_lockspace */
+#if PY_MAJOR_VERSION == 2
 PyDoc_STRVAR(pydoc_rem_lockspace, "\
-rem_lockspace(lockspace, host_id, path, offset=0, async=False, unused=False)\n\
-Remove a lockspace, releasing the acquired host_id. If async is True the\n\
+rem_lockspace(lockspace, host_id, path, offset=0, async=False, unused=False, wait=True)\n\
+Remove a lockspace, releasing the acquired host_id. If wait is False the\n\
+function will return immediately and the status can be checked using\n\
+inq_lockspace; This is identical to having the async flag set to True.\n\
+async flag is maintained only for backward compatibility and will be\n\
+deprecated in sanlock release for python 3.\n\
+If unused is True the command will fail (EBUSY) if there is\n\
+at least one acquired resource in the lockspace. Otherwise (the default)\n\
+sanlock will try to terminate processes holding resource leases and upon\n\
+successful termination these leases will be released.");
+#else
+PyDoc_STRVAR(pydoc_rem_lockspace, "\
+rem_lockspace(lockspace, host_id, path, offset=0, unused=False, wait=True)\n\
+Remove a lockspace, releasing the acquired host_id. If wait is False the\n\
 function will return immediately and the status can be checked using\n\
 inq_lockspace. If unused is True the command will fail (EBUSY) if there is\n\
 at least one acquired resource in the lockspace. Otherwise (the default)\n\
 sanlock will try to terminate processes holding resource leases and upon\n\
 successful termination these leases will be released.");
+#endif
 
 static PyObject *
 py_rem_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
 {
-    int rv = -1, async = 0, unused = 0, flags = 0;
-    PyObject *lockspace = NULL;
-    PyObject *path = NULL;
+    int rv = -1, flags = 0;
     struct sanlk_lockspace ls;
 
-    static char *kwlist[] = {"lockspace", "host_id", "path", "offset",
-                                "async", "unused", NULL};
-
-    /* initialize lockspace structure */
-    memset(&ls, 0, sizeof(struct sanlk_lockspace));
-
     /* parse python tuple */
-    if (!PyArg_ParseTupleAndKeywords(args, keywds, "O&kO&|kii", kwlist,
-        convert_to_pybytes, &lockspace, &ls.host_id, pypath_converter, &path,
-        &ls.host_id_disk.offset,
-        &async, &unused)) {
-        goto finally;
-    }
```

Comment:
Only the code above is relevant for argument parsing and handling wait/async.

Reply:
it is shared in rem_lockspace_parse_args()